src/server/container_utils.h

```diff
@@ -0,0 +1,169 @@
+quicklistEntry QLEntry() {
+  quicklistEntry res{.quicklist = NULL,
+                     .node = NULL,
+                     .zi = NULL,
+                     .value = NULL,
+                     .longval = 0,
+                     .sz = 0,
+                     .offset = 0};
+  return res;
+}
+
+// Iterate over all values and call func(val). Iteration stops as soon
+// as func return false.
+template <typename F> bool IterateList(PrimeIterator it, F&& func) {
+  quicklist* ql = static_cast<quicklist*>(it->second.RObjPtr());
+  quicklistIter* qiter = quicklistGetIteratorAtIdx(ql, AL_START_HEAD, 0);
+  quicklistEntry entry = QLEntry();
+  bool success = true;
+  while (success && quicklistNext(qiter, &entry)) {
+    if (entry.value) {
+      success &= func(std::string(reinterpret_cast<char*>(entry.value), entry.sz));
+    } else {
+      success &= func(entry.longval);
+    }
+  }
+  quicklistReleaseIterator(qiter);
+  return success;
+}
+
+// Iterate over all values and call func(val). If func returns a bool and
+// returns false, iteration stops. func might also return void.
+template <typename F, typename RetV = std::invoke_result_t<F, std::string&&>>
+RetV IterateSet(PrimeIterator it, F&& func) {
+  void* ptr = it->second.RObjPtr();
+  unsigned encoding = it->second.Encoding();
+
+  static_assert(std::is_same_v<RetV, void> || std::is_same_v<RetV, bool>,
+                "F needs to return void or bool");
+  auto wrapped_func = [func = std::forward<F>(func)](auto&& val) -> bool {
+    if constexpr (std::is_same_v<RetV, void>) {
+      func(std::forward<decltype(val)>(val));
+      return true;
+    } else {
+      return func(std::forward<decltype(val)>(val));
+    }
+  };
+
+  bool success = true;
+  if (encoding == kEncodingIntSet) {
+    intset* is = static_cast<intset*>(ptr);
+    int64_t ival;
+    int ii = 0;
+    char buf[32];
+
+    while (success && intsetGet(is, ii++, &ival)) {
+      if constexpr (std::is_invocable_v<F, int64_t>) {
+        success &= wrapped_func(ival);
+      } else {
+        char* next = absl::numbers_internal::FastIntToBuffer(ival, buf);
+        success &= wrapped_func(std::string{buf, size_t(next - buf)});
+      }
+    }
+  } else {
+    std::string str;
+    if (absl::GetFlag(FLAGS_use_set2)) {
+      for (sds ptr : *(StringSet*)ptr) {
+        str.assign(ptr, sdslen(ptr));
+        if (!wrapped_func(std::move(str))) {
+          success = false;
+          break;
+        }
+      }
+    } else {
+      dict* ds = (dict*)ptr;
+      dictIterator* di = dictGetIterator(ds);
+      dictEntry* de = nullptr;
+      while (success && (de = dictNext(di))) {
+        str.assign((sds)de->key, sdslen((sds)de->key));
+        success &= wrapped_func(std::move(str));
+      }
+      dictReleaseIterator(di);
+    }
+  }
+  return success;
+}
+
+// Iterate over all values and call func(val). Iteration stops as soon
+// as func returns false.
+template <typename F> bool Iterate(PrimeIterator it, F&& func) {
```

Comment:
you do not really use `Iterator`. Why not accept `const PrimeValue& obj` instead and call it with `it->second` ?